packages/hebao_v1/contracts/modules/dapps/CompoundModule.sol

```diff
@@ -85,13 +118,62 @@ contract CompoundModule is BaseSubAccount, SecurityModule
         address cToken = compoundRegistry.getCToken(token);
         require(cToken != address(0), "NO_MARKET");
 
-        uint balance = uint(tokenBalance(wallet, token));
-        require(amount > 0 && amount <= balance, "INVALID_WITHDRAW_AMOUNT");
+        require(amount > 0, "ZERO_WITHDRAW_AMOUNT");
+
+        if (isCollaterlCToken(wallet, cToken)) {
+            // for collaterl token, check liquidity to withdraw.
+            (uint collateralRemovable, uint borrowableAfterRemove) = collateralTokenRemovable(wallet, token);
+            require(amount <= collateralRemovable + borrowableAfterRemove, "INVALID_WITHDRAW_AMOUNT");
+
+            uint redeemable = collateralRemovable;
+            uint borrowable;
+            if (amount <= collateralRemovable) {
+                redeemable = amount;
+            } else {
+                borrowable = amount - collateralRemovable;
+            }
+
+            if (redeemable > 0) {
+                redeemUnderlying(wallet, cToken, redeemable);
+                trackWithdrawal(wallet, token, redeemable);
+            }
 
-        redeem(wallet, cToken, amount);
-        trackWithdrawal(wallet, token, amount);
+            if (borrowable > 0) {
+                borrow(wallet, cToken, borrowable);
+                trackBorrow(wallet, token, borrowable);
+            }
+            return;
+        }
+
+        // Non-collateral token, user exchange current cToken and borrow more according to liquidity.
+        // 2 cases:
+        //  a. amount <= balance                                  ==> redeem();
+        //  b. amount > balance && amount - balance <= borrowable ==> redeem() && borrow();
+        uint balance = CToken(cToken).balanceOfUnderlying(wallet);
+        uint borrowable = tokenBorrowable(wallet, token); // other collaterals tokens as this one is not.
+        require(amount <= balance + borrowable, "INVALID_WITHDRAW_AMOUNT");
+
+        uint withdrawAmount;
+        uint borrowAmount;
+        if (amount <= balance) {
+            withdrawAmount = amount;
+        } else {
+            withdrawAmount = balance;
+            borrowAmount = amount - withdrawAmount;
+        }
+
+        if (withdrawAmount > 0) {
+            redeemUnderlying(wallet, cToken, withdrawAmount);
+            trackWithdrawal(wallet, token, amount);
+        }
+
+        if (borrowAmount > 0) {
+            borrow(wallet, cToken, borrowAmount);
+            trackBorrow(wallet, token, borrowAmount);
+        }
     }
 
+    /// @dev tokenBalance is just the mapping from cToken to token, regardless of the collateral part.
```

Comment:
Shouldn't this return a negative number when the user has a loan open?

Reply:
you right, fixed.